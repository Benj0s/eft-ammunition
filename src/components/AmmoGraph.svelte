<script>
  import { onMount } from "svelte";
  import { scaleLinear } from "d3-scale";
  import { generateHexColourFromString } from "../services/colour-mapper.service";

  export let ammunition;
  export let cartridges;

  let svg;
  let width = 500;
  let height = 500;

  const padding = { top: 40, right: 40, bottom: 40, left: 40 };
  const maxDamage =
    Math.ceil(
      Math.max.apply(null, cartridges.map(cartridge => cartridge.damage)) / 10
    ) * 10;
  const maxPenetration =
    Math.ceil(
      Math.max.apply(null, cartridges.map(cartridge => cartridge.penetration)) /
        10
    ) * 10;
  const damageTicks = Array(maxDamage / 10 + 1)
    .fill()
    .map((a, i) => i * 10);
  const penetrationTicks = Array(maxPenetration / 10 + 1)
    .fill()
    .map((a, i) => i * 10);

  $: xScale = scaleLinear()
    .domain([0, maxDamage])
    .range([padding.left, width - padding.right]);

  $: yScale = scaleLinear()
    .domain([0, maxPenetration])
    .range([height - padding.bottom, padding.top]);

  $: xTicks = damageTicks;
  $: yTicks = penetrationTicks;

  onMount(resize);

  function resize() {
    ({ width, height } = svg.getBoundingClientRect());
  }
</script>

<style>
  .graph-wrapper {
    height: 100%;
    display: flex;
  }

  .graph {
    display: flex;
    flex: 1;
    height: 100%;
  }

  svg {
    width: 100%;
    height: 100%;
  }

  circle {
    fill-opacity: 0.6;
    stroke: rgba(0, 0, 0, 0.5);
  }

  .tick line {
    stroke: #ddd;
    stroke-dasharray: 2;
  }

  text {
    font-size: 12px;
    fill: #999;
  }

  .x-axis text {
    text-anchor: middle;
  }

  .y-axis text {
    text-anchor: end;
  }
</style>

<svelte:window on:resize={resize} />
<div class="graph-wrapper">
  <div class="graph">
    <svg bind:this={svg}>
      <!-- y axis -->
      <g class="axis y-axis">
        {#each yTicks as tick}
          <g class="tick tick-{tick}" transform="translate(0, {yScale(tick)})">
            <line x1={padding.left} x2={xScale(100)} />
            <text x={padding.left - 8} y="+4">{tick}</text>
          </g>
        {/each}
      </g>

      <!-- x axis -->
      <g class="axis x-axis">
        {#each xTicks as tick}
          <g class="tick" transform="translate({xScale(tick)},0)">
            <line y1={yScale(0)} y2={yScale(100)} />
            <text y={height - padding.bottom + 16}>{tick}</text>
          </g>
        {/each}
      </g>

      <!-- data -->
      {#each cartridges as cartridge}
        <circle
          cx={xScale(cartridge.damage)}
          cy={yScale(cartridge.penetration)}
          r="5"
          fill={generateHexColourFromString(cartridge.ammunitionId)} />
        <text
          x={xScale(cartridge.damage)}
          y={yScale(cartridge.penetration) + 20}
          text-anchor="middle"
          stroke="gray"
          stroke-width="1px">
          {cartridge.label}
        </text>
      {/each}
    </svg>
  </div>
</div>
