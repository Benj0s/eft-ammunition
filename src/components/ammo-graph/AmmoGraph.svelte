<script>
  import { onMount } from "svelte";
  import { scaleLinear } from "d3-scale";
  import { generateHexColourFromString } from "../../services/colour-mapper.service";
  import { maxAndPad, generateTicks } from "./AmmoGraph.service.js";
  import { displayedAmmo } from "../../store.js";

  export let cartridges;

  let selectedCartridges;

  let svg;
  let width = 500;
  let height = 500;

  const padding = { top: 40, right: 40, bottom: 40, left: 40 };
  let maxDamage = maxAndPad(cartridges.map(cartridge => cartridge.damage));
  let maxPen = maxAndPad(cartridges.map(cartridge => cartridge.penetration));

  displayedAmmo.subscribe(ammo => {
    selectedCartridges = cartridges.filter(
      cartridge => ammo.id === cartridge.ammunitionId
    );

    if (selectedCartridges.length) {
      maxDamage = maxAndPad(
        selectedCartridges.map(cartridge => cartridge.damage)
      );
      maxPen = maxAndPad(
        selectedCartridges.map(cartridge => cartridge.penetration)
      );
    }
  });

  $: xScale = scaleLinear()
    .domain([0, maxDamage])
    .range([padding.left, width - padding.right]);

  $: yScale = scaleLinear()
    .domain([0, maxPen])
    .range([height - padding.bottom, padding.top]);

  $: xTicks = generateTicks(maxDamage);
  $: yTicks = generateTicks(maxPen);

  onMount(resize);

  function resize() {
    ({ width, height } = svg.getBoundingClientRect());
  }
</script>

<style>
  .ammo-graph {
    padding: 6px;
    height: 100%;
    display: flex;
    background-color: #565656;
  }

  .graph {
    display: flex;
    flex: 1;
    height: 100%;
  }

  svg {
    width: 100%;
    height: 100%;
  }

  .tick line {
    stroke: #a4a4a4;
    stroke-dasharray: 2;
  }

  text {
    font-size: 12px;
    fill: #ddd;
  }

  .armour-class {
    fill: #999;
  }
  .x-axis text {
    text-anchor: middle;
  }

  .y-axis text {
    text-anchor: end;
  }

  .mark-circle {
    fill: #4e8cb6;
    stroke: #d3d3d3;
    opacity: 0.9;
  }

  .mark-text {
    stroke-width: 2px;
    fill: #ddd;
  }
</style>

<svelte:window on:resize={resize} />
<div class="ammo-graph">
  <div class="graph">
    <svg bind:this={svg}>
      <!-- y axis -->
      <g class="axis y-axis">
        <text
          text-anchor="middle"
          transform="translate(0, {xScale(maxPen) / 2}) rotate(90)">
          Penetration
        </text>
        {#each yTicks as tick, i}
          <g class="tick tick-{tick}" transform="translate(0, {yScale(tick)})">
            <line x1={padding.left} x2={xScale(maxDamage)} />
            <text x={padding.left - 8} y="+4">{tick}</text>
            {#if i > 0 && i < yTicks.length - 1}
              <text class="armour-class" x={padding.left + 90} y="-2">
                Armour class {tick / 10}
              </text>
            {/if}
          </g>
        {/each}
      </g>

      <!-- x axis -->
      <g class="axis x-axis">
        <text text-anchor="middle" x={xScale(maxDamage) / 2} y={height - 2}>
          Damage
        </text>
        {#each xTicks as tick}
          <g class="tick" transform="translate({xScale(tick)},0)">
            <line y1={yScale(0)} y2={yScale(maxPen)} />
            <text y={height - padding.bottom + 16}>{tick}</text>
          </g>
        {/each}
      </g>

      <!-- data -->
      {#each selectedCartridges as cartridge}
        <circle
          cx={xScale(cartridge.damage)}
          cy={yScale(cartridge.penetration)}
          r="5"
          class="mark-circle" />
        <text
          x={xScale(cartridge.damage)}
          y={yScale(cartridge.penetration) + 20}
          text-anchor="middle"
          class="mark-text">
          {cartridge.label}
        </text>
      {/each}
    </svg>
  </div>
</div>
