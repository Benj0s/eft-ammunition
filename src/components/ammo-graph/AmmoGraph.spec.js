import AmmoGraph from './AmmoGraph.svelte'
import { render } from '@testing-library/svelte'
// import { displayedAmmo, compareAllAmmo } from '../../store.js'

describe('AmmoGraph', () => {
  it('should render the component', async () => {
    const { container } = render(AmmoGraph, {
      props: {
        cartridges: [{ damage: 50, penetration: 60, label: 'ABC' }],
      },
    })

    expect(container).toMatchSnapshot()
  })
})
