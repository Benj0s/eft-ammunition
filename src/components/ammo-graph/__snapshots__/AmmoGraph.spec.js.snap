// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmmoGraph should render the component 1`] = `
<body>
  <div>
    <div
      class="ammo-graph svelte-ndcy92"
    >
      <div
        class="graph svelte-ndcy92"
      >
        <svg
          class="svelte-ndcy92"
        >
          <g
            class="axis y-axis svelte-ndcy92"
          >
            <text
              class="svelte-ndcy92"
              text-anchor="middle"
              transform="translate(0, 0) rotate(90)"
            >
              Penetration
        
            </text>
            <g
              class="tick tick-0 svelte-ndcy92"
              transform="translate(0, -40)"
            >
              <line
                class="svelte-ndcy92"
                x1="40"
                x2="-40"
              />
              <text
                class="svelte-ndcy92"
                x="36"
                y="+4"
              >
                0
              </text>
            </g>
            <g
              class="tick tick-10 svelte-ndcy92"
              transform="translate(0, -26.66666666666667)"
            >
              <line
                class="svelte-ndcy92"
                x1="40"
                x2="-40"
              />
              <text
                class="svelte-ndcy92"
                x="36"
                y="+4"
              >
                10
              </text>
              <text
                class="armour-class svelte-ndcy92"
                x="130"
                y="-2"
              >
                Armour class 
                1
              </text>
            </g>
            <g
              class="tick tick-20 svelte-ndcy92"
              transform="translate(0, -13.33333333333334)"
            >
              <line
                class="svelte-ndcy92"
                x1="40"
                x2="-40"
              />
              <text
                class="svelte-ndcy92"
                x="36"
                y="+4"
              >
                20
              </text>
              <text
                class="armour-class svelte-ndcy92"
                x="130"
                y="-2"
              >
                Armour class 
                2
              </text>
            </g>
            <g
              class="tick tick-30 svelte-ndcy92"
              transform="translate(0, 0)"
            >
              <line
                class="svelte-ndcy92"
                x1="40"
                x2="-40"
              />
              <text
                class="svelte-ndcy92"
                x="36"
                y="+4"
              >
                30
              </text>
              <text
                class="armour-class svelte-ndcy92"
                x="130"
                y="-2"
              >
                Armour class 
                3
              </text>
            </g>
            <g
              class="tick tick-40 svelte-ndcy92"
              transform="translate(0, 13.333333333333329)"
            >
              <line
                class="svelte-ndcy92"
                x1="40"
                x2="-40"
              />
              <text
                class="svelte-ndcy92"
                x="36"
                y="+4"
              >
                40
              </text>
              <text
                class="armour-class svelte-ndcy92"
                x="130"
                y="-2"
              >
                Armour class 
                4
              </text>
            </g>
            <g
              class="tick tick-50 svelte-ndcy92"
              transform="translate(0, 26.66666666666667)"
            >
              <line
                class="svelte-ndcy92"
                x1="40"
                x2="-40"
              />
              <text
                class="svelte-ndcy92"
                x="36"
                y="+4"
              >
                50
              </text>
              <text
                class="armour-class svelte-ndcy92"
                x="130"
                y="-2"
              >
                Armour class 
                5
              </text>
            </g>
            <g
              class="tick tick-60 svelte-ndcy92"
              transform="translate(0, 40)"
            >
              <line
                class="svelte-ndcy92"
                x1="40"
                x2="-40"
              />
              <text
                class="svelte-ndcy92"
                x="36"
                y="+4"
              >
                60
              </text>
            </g>
          </g>
          <g
            class="axis x-axis svelte-ndcy92"
          >
            <text
              class="svelte-ndcy92"
              text-anchor="middle"
              x="0"
              y="-2"
            >
              Damage
            </text>
            <g
              class="tick svelte-ndcy92"
              transform="translate(40,0)"
            >
              <line
                class="svelte-ndcy92"
                y1="-40"
                y2="40"
              />
              <text
                class="svelte-ndcy92"
                y="-22"
              >
                0
              </text>
            </g>
            <g
              class="tick svelte-ndcy92"
              transform="translate(24,0)"
            >
              <line
                class="svelte-ndcy92"
                y1="-40"
                y2="40"
              />
              <text
                class="svelte-ndcy92"
                y="-30"
              >
                10
              </text>
            </g>
            <g
              class="tick svelte-ndcy92"
              transform="translate(8,0)"
            >
              <line
                class="svelte-ndcy92"
                y1="-40"
                y2="40"
              />
              <text
                class="svelte-ndcy92"
                y="-22"
              >
                20
              </text>
            </g>
            <g
              class="tick svelte-ndcy92"
              transform="translate(-8,0)"
            >
              <line
                class="svelte-ndcy92"
                y1="-40"
                y2="40"
              />
              <text
                class="svelte-ndcy92"
                y="-30"
              >
                30
              </text>
            </g>
            <g
              class="tick svelte-ndcy92"
              transform="translate(-24,0)"
            >
              <line
                class="svelte-ndcy92"
                y1="-40"
                y2="40"
              />
              <text
                class="svelte-ndcy92"
                y="-22"
              >
                40
              </text>
            </g>
            <g
              class="tick svelte-ndcy92"
              transform="translate(-40,0)"
            >
              <line
                class="svelte-ndcy92"
                y1="-40"
                y2="40"
              />
              <text
                class="svelte-ndcy92"
                y="-30"
              >
                50
              </text>
            </g>
          </g>
          <circle
            class="mark-circle selected svelte-ndcy92"
            cx="-40"
            cy="40"
            r="5"
          />
          <text
            class="mark-text selected svelte-ndcy92"
            text-anchor="middle"
            x="-40"
            y="56"
          >
            ABC
          </text>
        </svg>
      </div>
    </div>
  </div>
</body>
`;
